### PR TITLE
Remove history after no-source outputs have been cleaned

### PR DESCRIPTION
**Pull request by wolfs**
So it is as if the task never executed.

### CODE REVIEW

**Comment by lptr** [comment]
I'd simplify this a lot by adding `ExecutionHistoryStore` in `SkipEmptySourceFilesTaskExecuter` and calling `remove(task.getPath())` directly there.

in #8038 I've gutted `TaskArtifactState`, doing the above would allow us to avoid conflicts, too.

**Comment by big-guy** [comment]
Is this a regression in 5.1?

**Comment by wolfs** [comment]
@lptr I started out injecting the execution history store directly in `SkipEmptySourceFilesTaskExecuter`. I ended up not doing it since it would require to calculate the key in the task history in two places, not only in `DefaultTaskArtifactStateRepository`.

I guess we can do it as you suggested, since it is easier and we will probably move this executor to somewhere else anyway.

**Comment by lptr** [comment]
Here's how I'm doing it in the PR now: https://github.com/gradle/gradle/pull/8038/commits/00b09500ab680d27af5ece7efefd828b95270ece

**Comment by wolfs** [comment]
@lptr PTAL!

File: subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java [hunk 1]
```diff
@@ -101,7 +104,7 @@ public TaskExecuterResult execute(TaskInternal task, TaskStateInternal state, fi
                 } else {
                     state.setOutcome(TaskExecutionOutcome.NO_SOURCE);
                 }
-                taskArtifactState.snapshotAfterTaskExecution(context);
+                executionHistoryStore.remove(task.getPath());
```
Comment (lptr) [inline subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java]: @wolfs I just realized that this should be one line lower, no? We should remove the history even if there were no output files during the previous execution.

Comment (lptr) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java]: I guess it's okay to leave it like this on `release`. I've moved it lower on `master`.

Comment (wolfs) [reply subprojects/core/src/main/java/org/gradle/api/internal/tasks/execution/SkipEmptySourceFilesTaskExecuter.java]: I guess such a task should always produce outputs when it has sources.

I would only remove the history if there is any, i.e. if `previousExecution != null`.

**Pull request merged**